### ISSUE TITLE
MySQL 8.0 Compatibility

### ISSUE THREAD

**Issue by morgo** [pingcap/tidb#7968]
This issue is to track what is required to move from MySQL 5.7 to [MySQL 8.0](https://dev.mysql.com/doc/refman/8.0/en/mysql-nutshell.html) compatibility.

## Essential

- [ ] Common Table Expressions (regular and recursive)
- [ ] Window Functions
- [ ] `caching_sha256_password` Authentication Plugin
- [ ] `utf8mb4` as default (rename `utf8` as in TiDB)
- [ ] Role Based Authentication
- [ ] Invisible Indexes

## Nice to Have
- [ ] `JSON_TABLE` function + add [remaining JSON functions](https://github.com/pingcap/tidb/issues/7546) that were backported to 5.7
- [ ] Regular expression support
- [ ] Cleanup [`SHOW VARIABLES`](https://dev.mysql.com/doc/refman/8.0/en/added-deprecated-removed.html) output (remove query cache etc.)
- [ ] Remove removed functions (insecure crypto etc)
- [ ] Maintain information about password history
- [ ] Resource Management
- [ ] Default values for `BLOB/TEXT` + default value as function
- [ ] GIS Support
- [ ] `SKIP LOCKED` and `NOWAIT` (I don't believe this is possible with optimistic locking?)

## Not Applicable
- New Data Dictionary
- Performance Schema enhancements
- Replication enhancements
- InnoDB enhancements

**Comment by shenli** [comment]
@morgo Nice work! We will add it to our roadmap.
/cc @winkyao @zz-jason @jackysp

**Comment by elvizlai** [comment]
is there any progress or plan about `JSON_TABLE` ?

**Comment by morgo** [comment]
I moved skip-locked to the essential list. At the time this list was created, the default was optimistic locking. But since it is now pessmistic, it is highly desirable.

**Comment by dveeden** [comment]
User attributes might also be something to add
```
mysql 8.0.22 > CREATE USER 'foo'@'%' IDENTIFIED BY 'affalfajljfjlk' COMMENT 'Support ticket #123';
Query OK, 0 rows affected (0.01 sec)

mysql 8.0.22 > CREATE USER 'bar'@'%' IDENTIFIED BY 'affalfajljfjlk' ATTRIBUTE '{"ticket": 123, "full_name": "John Doe"}';
Query OK, 0 rows affected (0.03 sec)
```

```
mysql 8.0.22 > select User_attributes from mysql.user where user in ('foo','bar');
+--------------------------------------------------------+
| User_attributes                                        |
+--------------------------------------------------------+
| {"metadata": {"ticket": 123, "full_name": "John Doe"}} |
| {"metadata": {"comment": "Support ticket #123"}}       |
+--------------------------------------------------------+
2 rows in set (0.01 sec)
```

**Comment by dveeden** [comment]
Should we add `utf8mb4_0900_ai_ci` to the list as well?

**Comment by yahonda** [comment]
Would you consider to add https://github.com/pingcap/tidb/issues/32118 to this list?
I think it needs supported because it blocks using MySQL Connector/J 8.

**Comment by dveeden** [comment]
> Would you consider to add #32118 to this list? I think it needs supported because it blocks using MySQL Connector/J 8.

I don't think that really should block Connector/J 8.0. However I think fixing that soon-ish would be good anyway.

**Comment by dveeden** [comment]
Maybe we should add https://github.com/pingcap/dm/issues/1950 to this list? or is this strictly for tidb-server?

**Comment by bb7133** [comment]
> Would you consider to add https://github.com/pingcap/tidb/issues/32118 to this list?

@yahonda @dveeden , I think this is a 'nice to have' compatibility improvement. Personally I'm really looking forward to close it, thanks, Daniel!

**Comment by SunRunAway** [comment]
Should the full implementation of [Dynamic Privileges](https://dev.mysql.com/doc/refman/8.0/en/privileges-provided.html) be essential?

**Comment by dveeden** [comment]
> Should the full implementation of [Dynamic Privileges](https://dev.mysql.com/doc/refman/8.0/en/privileges-provided.html) be essential?

I don't think that is essential.

**Comment by jaggerwang** [comment]
Multi-valued indexes on json arrays and descending indexes are exactly what we need, is there any plan for these two features?

**Comment by SunRunAway** [comment]
@jaggerwang 
AFAIK, Multi-valued indexes are planned in early next year /cc @bb7133 
For descending indexes, could you describe your scenarios?

**Comment by jaggerwang** [comment]
For descending indexes, suppose we have a table named `file` to store a filesystem's meta info.

| Name  | Type | Desc |
| ------------- | ------------- | ------------- |
| id | int | ID |
| type | int | 0 for directory, 1 for file |
| filename | string | File or directory name |
| created_at | datetime | Create time |

We have the following sql to query file list order by filename ASC or DESC, directories should before files.

```sql
# Order by filename ASC
select * from file order by type asc, filename asc limit 100;

# Order by filename DESC
select * from file order by type asc, filename desc limit 100;
```

To accelerate query, we created an union index `(type, filename)`, but the index is not working well when order by filename DESC, about 2 to 3 times slower.

**Comment by dveeden** [comment]
Improvements in password management are tracked in https://github.com/pingcap/tidb/issues/38923

**Comment by yahonda** [comment]
I've found `last_insert_id()` behavior differences between MySQL 5.7 and 8.0. TiDB 6.6.0 behaves same as MySQL 5.7. #42084

**Comment by yahonda** [comment]
It looks like TiDB does not support this feature available since MySQL 8.0.13 yet. This checkbox is checked somehow.

Default values for BLOB/TEXT + default value as function (8.0.13) https://github.com/pingcap/tidb/issues/10377

This example comes from the https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html
* TiDB 7.5.0
```sql
my:root@127.0.0.1:4000=> use test;
USE
my:root@127.0.0.1:4000=> CREATE TABLE t2 (b BLOB DEFAULT ('abc'));
error: mysql: 1064: You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 38 near "'abc'));"
```

```sql
my:root@127.0.0.1:4000=> select tidb_version();
                      tidb_version()
-----------------------------------------------------------
 Release Version: v7.5.0                                  +
 Edition: Community                                       +
 Git Commit Hash: 069631e2ecfedc000ffb92c67207bea81380f020+
 Git Branch: heads/refs/tags/v7.5.0                       +
 UTC Build Time: 2023-11-24 08:41:52                      +
 GoVersion: go1.21.3                                      +
 Race Enabled: false                                      +
 Check Table Before Drop: false                           +
 Store: tikv
(1 row)

my:root@127.0.0.1:4000=>
```

* MySQL 8.2.0
```sql
mysql> CREATE TABLE t2 (b BLOB DEFAULT ('abc'));
Query OK, 0 rows affected (0.12 sec)

mysql> show create table t2;
+-------+---------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                              |
+-------+---------------------------------------------------------------------------------------------------------------------------+
| t2    | CREATE TABLE `t2` (
  `b` blob DEFAULT (_utf8mb4'abc')
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci |
+-------+---------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> insert into t2 values;
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '' at line 1
mysql> insert into t2 values();
Query OK, 1 row affected (0.01 sec)

mysql> select * from t2;
+------------+
| b          |
+------------+
| 0x616263   |
+------------+
1 row in set (0.00 sec)

mysql> select version();
+-----------+
| version() |
+-----------+
| 8.2.0     |
+-----------+
1 row in set (0.01 sec)

mysql>
```

**Comment by yahonda** [comment]
It is not a big deal though `utf8mb4_0900_bin` collation has been introduced since MySQL 8.0.17. Here it is listed as "8.0.11 and before"

https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-17.html
https://github.com/mysql/mysql-server/commit/c7934d119aeb

**Comment by dveeden** [comment]
> It is not a big deal though `utf8mb4_0900_bin` collation has been introduced since MySQL 8.0.17. Here it is listed as "8.0.11 and before"
> 
> https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-17.html [mysql/mysql-server@c7934d119aeb](https://github.com/mysql/mysql-server/commit/c7934d119aeb)

Yes indeed. I've fixed that.

**Comment by dveeden** [comment]
> It looks like TiDB does not support this feature available since MySQL 8.0.13 yet. This checkbox is checked somehow.
> 
> Default values for BLOB/TEXT + default value as function (8.0.13) #10377
> 


Yes it looks like this was marked as completed after only part of this had been implemented. I think https://github.com/pingcap/tidb/issues/45506 covers the second part. I've update the description to reflect this.
